numba_cuda/numba/cuda/cudadrv/managed_module.py

```diff
@@ -0,0 +1,93 @@
+import weakref
+
+from . import devices
+from .driver import CtypesModule, USE_NV_BINDING
+
+
+class _CuFuncProxy:
+    """See `ManagedModule` for more details
+    """
+    def __init__(self, module, cufunc):
+        self._module = module
+        self._cufunc = cufunc
+
+    def init_module(self, stream):
+        self._module._init(stream)
+
+    def lazy_finalize_module(self, stream):
+        self._module._lazy_finalize(stream)
+
+    def __getattr__(self, name):
+        return getattr(self._cufunc, name)
+
+
+class ManagedModule:
+    def __init__(self, module, setup_callbacks, teardown_callbacks):
+        """ctypes module with setup and teardown callbacks
+        The use of managedmodule is the same as a ctypes module,
+        with the exception of `get_function`, which returns a wrapped
+        cufunc object. The wrapped object provides `init_module`
+        and `lazy_finalize_module` method. Which are used
+        to initialize and finalize the module when stream is available
+        in the later stage of the compilation pipeline.
+        """
+        # To be updated to object code
+        if not isinstance(module, CtypesModule):
+            raise TypeError("module must be a CtypesModule")
+
+        if not isinstance(setup_callbacks, list):
+            raise TypeError("setup_callbacks must be a list")
+        if not isinstance(teardown_callbacks, list):
+            raise TypeError("teardown_callbacks must be a list")
+
+        for callback in setup_callbacks:
+            if not callable(callback):
+                raise TypeError("all callbacks must be callable")
+        for callback in teardown_callbacks:
+            if not callable(callback):
+                raise TypeError("all callbacks must be callable")
+
+        self._initialized = False
+        self._module = module
+        self._setup_callbacks = setup_callbacks
+        self._teardown_callbacks = teardown_callbacks
+
+    def _init(self, stream):
+        """Eagerly call the setup functions for cumodule on `stream`
```

Comment:
I think the setup function isn't really being called on the stream - it's being called, and given a stream to use for CUDA operations. Is that correct?

Reply:
Correct - at this stage we don't know what operation the user is going to perform inside the callbacks.

Reply:
Thanks - in that case, would this more closely reflect what `_init()` does

```suggestion
        """Call the setup functions for cumodule with the given `stream`
```

Reply:
This is applied to the similar method in `Module` named `setup`.